### ISSUE TITLE
question on train_loss calculation mismatch

### ISSUE THREAD

**Issue by sushmit86** [skorch-dev/skorch#822]
Hi,
I noticed there is a mismatch between train_loss reported during a model fit vs if I calculate it later. Here is the code to recreate

```
##
import numpy as np
import torch
from torch import nn
import torch.nn.functional as F

import skorch
from skorch import NeuralNetClassifier
from skorch.dataset import Dataset
from skorch.helper import predefined_split
from sklearn.model_selection import train_test_split
from skorch.callbacks import EpochScoring
from sklearn.datasets import make_regression
X_regr, y_regr = make_regression(1000, 20, n_informative=10, random_state=0)
X_regr = X_regr.astype(np.float32)
y_regr = y_regr.astype(np.float32) / 100
y_regr = y_regr.reshape(-1, 1)

print(X_regr.shape, y_regr.shape, y_regr.min(), y_regr.max())

class RegressorModule(nn.Module):
    def __init__(
            self,
            num_units=10,
            nonlin=F.relu,
    ):
        super(RegressorModule, self).__init__()
        self.num_units = num_units
        self.nonlin = nonlin

        self.dense0 = nn.Linear(20, num_units)
        self.nonlin = nonlin
        self.dense1 = nn.Linear(num_units, 10)
        self.output = nn.Linear(10, 1)

    def forward(self, X, **kwargs):
        X = self.nonlin(self.dense0(X))
        X = F.relu(self.dense1(X))
        X = self.output(X)
        return X
from skorch import NeuralNetRegressor
net_regr = NeuralNetRegressor(
    RegressorModule,
    max_epochs=10,
    lr=0.1,
    device='cuda',  # uncomment this to train with CUDA
    train_split=predefined_split(Dataset(X_regr,y_regr))
)
net_regr.fit(X_regr, y_regr)
```
if I run the above code I notice the validation loss reported is different from train loss. Since train set = validation set I was expecting the two losses reported would be the same. To take a step further I calculated the loss using skleran metrics. Here is the code.
```

from sklearn.metrics import mean_squared_error
mean_squared_error(y_regr, net_regr.predict(X_regr))
```
The above exactly matches with validation_loss.Is the train_loss reported is just for a batch ? Not sure what I am missing here?

**Comment by BenjaminBossan** [comment]
The way the train loss is calculated is that the loss is stored after each individual batch and then averaged. So e.g. if you have 1000 samples and batch size of 100, you get 10 train losses. After each batch, the model is updated and thus has learned a little bit more. _After_ the 10 updates based on the training data, the loss on the validation data is measured, i.e. at a time when the model was already trained on 1000 samples. Therefore, the loss on the validation set will be different (in your case: better) than the one reported on the train set.

(This is also why you can sometimes observe that the model reports a better score on the validation data than on the training data, even though one should expect the opposite because of overfitting.)

I hope that explains both of your findings.

**Comment by sushmit86** [comment]
Thanks so much for your help. Had a couple of follow-ups if you can help. So if we run the above code without the training set. This is what we get as the last epoch result.

```
{'batches': [{'train_loss': 0.0955338403582573, 'train_batch_size': 128},
  {'train_loss': 0.08702770620584488, 'train_batch_size': 128},
  {'train_loss': 0.09187590330839157, 'train_batch_size': 128},
  {'train_loss': 0.06455320119857788, 'train_batch_size': 128},
  {'train_loss': 0.042118728160858154, 'train_batch_size': 128},
  {'train_loss': 0.03709245100617409, 'train_batch_size': 128},
  {'train_loss': 0.04171159490942955, 'train_batch_size': 128},
  {'train_loss': 0.0303038302809, 'train_batch_size': 104}],
 'epoch': 10,
 'train_batch_count': 8,
 'dur': 0.01835775375366211,
 'train_loss': 0.062020516768097875,
 'train_loss_best': True}


```
So the ```train_loss = 0.062020516768097875,``` is the average of all train_loss in batches?

Also was there a particular reason to implement the training loss in this way? ( I get it is computationally faster). if I look at the native PyTorch code below, it seems like train loss is calculated after the epoch. Do other deep learning frameworks like Keras follow the same approach ? 
```

num_epochs = 3
for epoch in range(num_epochs):
    for X, y in data_iter:
        l = loss(net(X), y)
        trainer.zero_grad()
        l.backward()
        trainer.step()
    l = loss(net(features), labels)
    print(f'epoch {epoch + 1}, loss {l:f}')
```

**Comment by BenjaminBossan** [comment]
> So the `train_loss = 0.062020516768097875,` is the average of all train_loss in batches?

Yes, weighted by the batch size.

> Also was there a particular reason to implement the training loss in this way? ( I get it is computationally faster).

As you mention, this is done for efficiency. In the code you provide, assuming that `features` are the training data, you would do the forward pass twice on the training data, slowing down the training process considerably. If you want to, you could add a callback in skorch that calculates the train loss after the parameter updates as an additional loss.

> Do other deep learning frameworks like Keras follow the same approach ?

I'm not sure how keras does it but probably it's done in a very similar fashion.

**Comment by sushmit86** [comment]
Thanks so much for the reply.

**Comment by BenjaminBossan** [comment]
I assume this is solved now. If not, feel free to re-open.
